File: lib/core_services/comic/mixin/comic_comment_mixin.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../../mixin/export.dart';
import '../main.dart';

part 'comic_comment_mixin.freezed.dart';
part 'comic_comment_mixin.g.dart';

@freezed
sealed class ComicCommentContext with _$ComicCommentContext {
  const factory ComicCommentContext({
    required String comicId,
    required MetaComic metaComic,
    String? chapterId,
    ComicChapter? chapter,
    ComicComment? parent,
  }) = _ComicCommentContext;

  factory ComicCommentContext.fromJson(Map<String, dynamic> json) =>
      _$ComicCommentContextFromJson(json);
}

mixin ComicCommentMixin on AuthMixin {
  // for comment system
  Future<ComicComments> getComments(ComicCommentContext context, {int? page});

  Future<void> deleteComment(
    ComicCommentContext context, {
    required ComicComment comment,
  });

  Future<bool> setLikeComment(
    ComicCommentContext context, {
    required ComicComment comment,
    required bool value,
  }) {
    throw UnimplementedError();
  }
}

File: lib/core_services/comic/mixin/comic_comment_mixin.ts
import type { MetaComic } from "../interfaces/meta_comic"
import type { ComicChapter } from "../interfaces/comic_chapter"
import type { ComicComment } from "../interfaces/comic_comment"
import type { ComicComments } from "../interfaces/comic_comments"
import { ComicAuthMixin } from "./comic_auth_mixin"

export interface ComicCommentContext {
  comicId: string
  metaComic: MetaComic
  chapterId?: string
  chapter?: ComicChapter
  parent?: ComicComment
}

export abstract class ComicCommentMixin extends ComicAuthMixin {
  // for comment system
  abstract getComments(
    context: ComicCommentContext,
    page?: number
  ): Promise<ComicComments>

  abstract deleteComment(
    context: ComicCommentContext,
    comment: ComicComment
  ): Promise<void>

  setLikeComment(
    context: ComicCommentContext,
    comment: ComicComment,
    value: boolean
  ): Promise<boolean> {
    throw new Error("UnimplementedError")
  }
}
